File: Maps3DSamples/ApiDemos/java-app/src/main/java/com/example/maps3djava/popovers/PopoversActivity.java
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.maps3djava.popovers;

import android.graphics.Color;
import android.graphics.Point;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.example.maps3djava.sampleactivity.SampleBaseActivity;
import com.google.android.gms.maps3d.GoogleMap3D;
import com.google.android.gms.maps3d.Popover;
import com.google.android.gms.maps3d.model.AltitudeMode;
import com.google.android.gms.maps3d.model.Camera;
import com.google.android.gms.maps3d.model.CollisionBehavior;
import com.google.android.gms.maps3d.model.LatLngAltitude;
import com.google.android.gms.maps3d.model.Map3DMode;
import com.google.android.gms.maps3d.model.Marker;
import com.google.android.gms.maps3d.model.MarkerOptions;
import com.google.android.gms.maps3d.model.PopoverOptions;
import com.google.android.gms.maps3d.model.PopoverShadow;
import com.google.android.gms.maps3d.model.PopoverStyle;

public class PopoversActivity extends SampleBaseActivity {

  private static final double CONTENT_LAT = 37.820642;
  private static final double CONTENT_LNG = -122.478227;
  private static final double CONTENT_ALT = 0.0;

  private Popover popover = null;
  private int popoverToggleCount = 0;

  @NonNull
  @Override
  public String getTAG() {
    return "PopoversActivity";
  }

  @NonNull
  @Override
  public Camera getInitialCamera() {
    return com.example.maps3d.common.UtilitiesKt.toValidCamera(new Camera(
        new LatLngAltitude(CONTENT_LAT, CONTENT_LNG, CONTENT_ALT),
        0.0,
        45.0,
        0.0,
        4075.0));
  }

  @Override
  public void onMap3DViewReady(GoogleMap3D googleMap3D) {
    super.onMap3DViewReady(googleMap3D);
    googleMap3D.setOnMapReadyListener((map) -> {
      googleMap3D.setOnMapReadyListener(null);
      onMapReady(googleMap3D);
    });
  }

  private void onMapReady(@NonNull GoogleMap3D googleMap3D) {
    googleMap3D.setMapMode(Map3DMode.SATELLITE);
    setupPopover(googleMap3D);
  }

  public void setupPopover(GoogleMap3D googleMap3D) {
    LatLngAltitude center = new LatLngAltitude(37.819852, -122.478549, 0.0);

    MarkerOptions markerOptions = new MarkerOptions();
    markerOptions.setPosition(center);
    markerOptions.setLabel("Golden Gate Bridge");
    markerOptions.setZIndex(1);
    markerOptions.setExtruded(true);
    markerOptions.setDrawnWhenOccluded(true);
    markerOptions.setCollisionBehavior(CollisionBehavior.REQUIRED);
    markerOptions.setAltitudeMode(AltitudeMode.RELATIVE_TO_MESH);

    Marker markerInGoldenGate = googleMap3D.addMarker(markerOptions);

    if (markerInGoldenGate == null) {
      Log.e(getTAG(), "Failed to create marker");
      return;
    } else {
      Log.w(getTAG(), "Marker created");
    }

    PopoverShadow shadow = new PopoverShadow();
    shadow.setColor(Color.argb(77, 0, 0, 0)); // 0.3 * 255 = 76.5 ~= 77
    shadow.setOffsetX(2.0f);
    shadow.setOffsetY(4.0f);
    shadow.setRadius(4.0f);

    PopoverStyle style = new PopoverStyle();
    style.setPadding(20.0f);
    style.setBackgroundColor(Color.WHITE);
    style.setBorderRadius(8.0f);
    style.setShadow(shadow);

    PopoverOptions popoverOptions = new PopoverOptions();
    popoverOptions.setContent(createGoldenGateInfoView());
    popoverOptions.setPositionAnchor(markerInGoldenGate);
    popoverOptions.setAutoPanEnabled(true);
    popoverOptions.setAutoCloseEnabled(true);
    popoverOptions.setAnchorOffset(new Point(0, 0));
    popoverOptions.setPopoverStyle(style);

    popover = googleMap3D.addPopover(popoverOptions);

    markerInGoldenGate.setClickListener(() -> {
      Log.d(getTAG(), "Marker clicked");
      if (popoverToggleCount > 5) {
        runOnUiThread(() -> {
          if (popover != null) {
            popover.remove();
          }
        });
        Log.d(getTAG(), "Popover removed");
        popoverToggleCount = 0;
      } else {
        Log.d(getTAG(), "Popover toggled");
        runOnUiThread(() -> {
          if (popover != null) {
            popover.toggle();
          }
        });
        popoverToggleCount++;
      }
    });

    Log.d(getTAG(), "Popover created");
  }

  private View createGoldenGateInfoView() {
    LinearLayout layout = new LinearLayout(this);
    layout.setOrientation(LinearLayout.VERTICAL);
    layout.setLayoutParams(new LinearLayout.LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));

    TextView titleView = new TextView(this);
    titleView.setText("The Golden Gate Bridge");
    titleView.setTextSize(18f);
    titleView.setTextColor(Color.BLACK);
    layout.addView(titleView);

    TextView headlineView = new TextView(this);
    headlineView.setText("San Francisco, CA");
    headlineView.setTextSize(14f);
    headlineView.setTextColor(Color.DKGRAY);
    layout.addView(headlineView);

    TextView contentView = new TextView(this);
    contentView.setText("The Golden Gate Bridge is a suspension bridge\n" +
        " spanning the one-mile-wide strait connecting\n" +
        " San Francisco Bay and the Pacific Ocean.\n" +
        " The bridge was completed in 1937.");
    contentView.setTextSize(12f);
    contentView.setTextColor(Color.GRAY);
    layout.addView(contentView);

    return layout;
  }
}

File: Maps3DSamples/ApiDemos/kotlin-app/src/main/java/com/example/maps3dkotlin/popovers/PopoversActivity.kt
package com.example.maps3dkotlin.popovers

import android.graphics.Color
import android.graphics.Point
import android.util.Log
import android.view.View
import android.view.ViewGroup.LayoutParams
import android.widget.LinearLayout
import android.widget.TextView
import com.example.maps3dkotlin.sampleactivity.SampleBaseActivity
import com.google.android.gms.maps3d.GoogleMap3D
import com.google.android.gms.maps3d.Popover
import com.google.android.gms.maps3d.model.AltitudeMode
import com.google.android.gms.maps3d.model.CollisionBehavior
import com.google.android.gms.maps3d.model.Map3DMode
import com.google.android.gms.maps3d.model.camera
import com.google.android.gms.maps3d.model.latLngAltitude
import com.google.android.gms.maps3d.model.markerOptions
import com.google.android.gms.maps3d.model.popoverOptions
import com.google.android.gms.maps3d.model.popoverShadow
import com.google.android.gms.maps3d.model.popoverStyle

class PopoversActivity : SampleBaseActivity() {
    override val TAG = "PopoversActivity"

    // San Francisco
    private val CONTENT_LAT = 37.820642
    private val CONTENT_LNG = -122.478227
    private val CONTENT_ALT = 0.0

    private var popover: Popover? = null
    private var popoverToggleCount = 0

    override val initialCamera = camera {
        center = latLngAltitude {
            latitude = CONTENT_LAT
            longitude = CONTENT_LNG
            altitude = CONTENT_ALT
        }
        heading = 0.0
        tilt = 45.0
        range = 4075.0
    }

    override fun onMapReady(googleMap3D: GoogleMap3D) {
        super.onMapReady(googleMap3D)
        googleMap3D.setMapMode(Map3DMode.SATELLITE)
        setupPopover(googleMap3D)
    }

    fun setupPopover(googleMap3D: GoogleMap3D) {
        val center = latLngAltitude {
            latitude = 37.819852
            longitude = -122.478549
            altitude = 0.0
        }

        val markerInGoldenGate =
            googleMap3D.addMarker(
                markerOptions {
                    position = center
                    label = "Golden Gate Bridge"
                    zIndex = 1
                    isExtruded = true
                    isDrawnWhenOccluded = true
                    collisionBehavior = CollisionBehavior.REQUIRED
                    altitudeMode = AltitudeMode.RELATIVE_TO_MESH
                }
            )

        if (markerInGoldenGate == null) {
            Log.e(TAG, "Failed to create marker")
            return
        } else {
            Log.w(TAG, "Marker created")
        }

        val popoverOptions = popoverOptions {
            content = createGoldenGateInfoView()
            positionAnchor = markerInGoldenGate
            autoPanEnabled = true
            autoCloseEnabled = true

            anchorOffset = Point(0, 0)
            popoverStyle = popoverStyle {
                padding = 20.0f
                backgroundColor = Color.WHITE
                borderRadius = 8.0f
                shadow = popoverShadow {
                    color = Color.argb(77, 0, 0, 0) // 0.3 * 255 = 76.5 ~= 77
                    offsetX = 2.0f
                    offsetY = 4.0f
                    radius = 4.0f
                }
            }
        }
        popover = googleMap3D.addPopover(popoverOptions)
        markerInGoldenGate.let {
            it.setClickListener {
                Log.d(TAG, "Marker clicked")
                if (popoverToggleCount > 5) {
                    runOnUiThread { popover?.remove() }
                    Log.d(TAG, "Popover removed")
                    popoverToggleCount = 0
                } else {
                    Log.d(TAG, "Popover toggled")
                    runOnUiThread { popover?.toggle() }
                    popoverToggleCount++
                }
            }
        }

        Log.d(TAG, "Popover created")
    }

    private fun createGoldenGateInfoView(): View {
        val context = this
        val layout =
            LinearLayout(context).apply {
                orientation = LinearLayout.VERTICAL
                layoutParams =
                    LinearLayout.LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT)
            }

        val titleView =
            TextView(context).apply {
                text = "The Golden Gate Bridge"
                textSize = 18f
                setTextColor(Color.BLACK)
            }
        layout.addView(titleView)

        val headlineView =
            TextView(context).apply {
                text = "San Francisco, CA"
                textSize = 14f
                setTextColor(Color.DKGRAY)
            }
        layout.addView(headlineView)

        val contentView =
            TextView(context).apply {
                text =
                    "The Golden Gate Bridge is a suspension bridge\n" +
                            " spanning the one-mile-wide strait connecting\n" +
                            " San Francisco Bay and the Pacific Ocean.\n" +
                            " The bridge was completed in 1937."
                textSize = 12f
                setTextColor(Color.GRAY)
            }
        layout.addView(contentView)

        return layout
    }

}
